src/lib/device-online-state.ts

```diff
@@ -0,0 +1,328 @@
+	private static readonly REDIS_NAMESPACE = 'device-online-state';
+	private static readonly EXPIRED_QUEUE = 'expired';
+	private static readonly RSMQ_READ_TIMEOUT = 30;
+
+	rsmq: RedisSMQ;
+	redis: PromisifedRedisClient;
+
+	constructor() {
+		// create a new Redis client...
+		this.redis = createPromisifedRedisClient({
+			host: REDIS_HOST,
+			port: REDIS_PORT,
+		});
+
+		// initialise the RedisSMQ object using our Redis client...
+		this.rsmq = new RedisSMQ({
+			client: this.redis,
+			ns: DeviceOnlineStateManager.REDIS_NAMESPACE,
+		});
+
+		// create the RedisMQ queue and start consuming messages...
+		this.rsmq
+			.createQueueAsync({ qname: DeviceOnlineStateManager.EXPIRED_QUEUE })
+			.catch(err => {
+				if (err.name !== 'queueExists') {
+					throw err;
+				}
+			})
+			.then(() => {
+				this.consume();
+			});
+	}
+
+	private updateDeviceModel(uuid: string, newState: DeviceOnlineStates) {
+		console.log(
+			`DeviceStateManager: ${uuid} is ${DeviceOnlineStates[newState]}`,
+		);
+
+		// patch the api_heartbeat_state value to the new state...
+		resinApi
+			.patch({
+				resource: 'device',
+				options: {
+					$filter: {
+						uuid,
+						api_heartbeat_state: {
+							$ne: newState,
+						},
+					},
+				},
+				body: {
+					api_heartbeat_state: newState,
+				},
+				passthrough: {
+					req: root,
+				},
+			})
+			.catch(err => {
+				captureException(
+					err,
+					'DeviceStateManager: Error updating the API with the device new state.',
+				);
+			});
+	}
+
+	// event emitters
+	private deviceOnline(uuid: string) {
+		this.updateDeviceModel(uuid, DeviceOnlineStates.Online);
+	}
+
+	private deviceOffline(uuid: string) {
+		this.updateDeviceModel(uuid, DeviceOnlineStates.Offline);
+	}
+
+	private deviceTimeout(uuid: string) {
+		this.updateDeviceModel(uuid, DeviceOnlineStates.Timeout);
+	}
+
+	private consume() {
+		// pull a message from the queue...
+		this.rsmq
+			.receiveMessageAsync({
+				qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+				vt: DeviceOnlineStateManager.RSMQ_READ_TIMEOUT, // prevent other consumers seeing the same message (if any) preventing multiple API agents from processing it...
+			})
+			.then(msg => {
+				if ('id' in msg) {
+					const { id, message } = msg;
+
+					return Bluebird.try(() => {
+						const { uuid, nextState } = JSON.parse(message) as {
+							uuid: string;
+							nextState: DeviceOnlineStates;
+						};
+
+						// raise and event for the state change...
+						switch (nextState) {
+							case DeviceOnlineStates.Timeout:
+								this.scheduleChangeOfStateForDevice(
+									uuid,
+									DeviceOnlineStates.Timeout,
+									DeviceOnlineStates.Offline,
+									API_HEARTBEAT_STATE_TIMEOUT_SECONDS, // put the device into a timeout state if it misses it's scheduled heartbeat window... then mark as offline
+								);
+								this.deviceTimeout(uuid);
+								break;
+							case DeviceOnlineStates.Offline:
+								this.deviceOffline(uuid);
```

Comment:
These generate a promise right?  Should this be
```suggestion
								return this.deviceOffline(uuid);
```
?
If not then you'll need to `return null` to avoid a bluebird warning

Reply:
I refactored this